### PR TITLE
doc: overhaul sample criteria and expectations

### PR DESCRIPTION
**Pull request by nashif**
overhaul sample criteria and what we expect in a sample, especially
overhaul the twister part and what should be provided for good test
coverage.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

### CODE REVIEW

File: samples/sample_definition_and_criteria.rst [hunk 1]
```diff
@@ -39,36 +39,43 @@ Sample Criteria
...
+    execution. Twister only attempts to flash and execute the build binary if
+    the harness is set to ztest or console.
+  * The default configuration for the sample should be in the ``prj.conf``
+    file and should be operational on all supported platforms. Do not rely on the
+    ``extra_args`` or ``extra_configs`` options in the YAML file to build the
+    sample.
+  * The tests should verify the default configuration of the sample and any
+    optional features documented in the sample's README file. Sample should
+    never be used to test functionality of the subsystem or module the sample
+    belongs to.
```
Comment (jukkar) [inline samples/sample_definition_and_criteria.rst:63]: > Sample should never be used to test functionality of the subsystem or module the sample belongs to.

Don't quite understand this, isn't the purpose of the various samples to test/show functionality of the subsystem, at least in networking samples do that.

Comment (nashif) [reply samples/sample_definition_and_criteria.rst:63]: show functionality: yes, act as a test: no.
Obviously when you run a sample you test something, but this should not replace actual tests with full coverage under tests/.
In a sample, the only thing you test is buildability and in some cases if a sample performing as expected, i.e. you are testing the sample, not the subsystem.

Comment (jukkar) [reply samples/sample_definition_and_criteria.rst:63]: Sure, I agree fully. It was just not clear from the text what it actually means. Perhaps some of your clarification text could be added to the document.

Comment (nashif) [reply samples/sample_definition_and_criteria.rst:70]: @jukkar  done

File: samples/sample_definition_and_criteria.rst [hunk 2]
```diff
@@ -11,11 +11,13 @@ more features, subsystems, or modules. This includes kernel services, drivers, p
 Samples should be limited in scope and should focus only on demonstrating non-trivial or
 essential aspects of the subsystem(s) or module(s) being highlighted.
 
+Samples are recommended when submitting new features, however they are not mandatory.
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:14]:
```suggestion
Samples are recommended when submitting new features; however, they are not mandatory.
```

File: samples/sample_definition_and_criteria.rst [hunk 3]
```diff
@@ -25,7 +27,13 @@ Sample Criteria
   * If a sample can provide output that can be verified, then output should be evaluated against
     expected value to have some level of testing for the sample itself.
     Refer to :ref:`twister_script` for more details.
-  * Samples are optional.
+
+  * Although being able to run a sample successfully does verify the
+    functionality is working as expected, this should not replace dedicated and
+    comprehensive tests with full coverage to be submitted into the
+    :zephyr_file:`tests/` folder.  In a sample, the only thing you test is
+    buildability and in some cases if a sample performing as expected, i.e. you
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:35]:
```suggestion
    buildability and, in some cases, if a sample is performing as expected, i.e. you
```

File: samples/sample_definition_and_criteria.rst [hunk 4]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
-      in ``samples/subsys/logging/syst``.
-    * Sample output can be validated leveraging the ``harness_config`` regex option,
-      wherever applicable.
-
-3. Samples should target multiple platforms and architectures.
-++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-  * Minimize the use of ``platform_allow`` and architecture filters as it limits the scope
-    of testing to the mentioned platforms and architectures.
-    Reference: :ref:`twister_script`
+  * Do not mark the test as build only. A sample should be able to build and
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:59]:
```suggestion
  * Do not mark the test as build only. A sample should be able to build *and*
```

File: samples/sample_definition_and_criteria.rst [hunk 5]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    run on the documented platforms. Use the harness option to skip the
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:60]:
```suggestion
    run on the documented platforms. Use the ``harness`` option to skip the
```

File: samples/sample_definition_and_criteria.rst [hunk 6]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    execution. Twister only attempts to flash and execute the build binary if
+    the harness is set to ztest or console.
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:62]:
```suggestion
    the harness is set to ``ztest`` or ``console``.
```

File: samples/sample_definition_and_criteria.rst [hunk 7]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+  * The default configuration for the sample should be in the ``prj.conf``
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:63]:
```suggestion
  * The default configuration for the sample must be in the :file:`prj.conf`
```

Also, as a general comment: maybe all the occurences of SHOULD / MUST / DO NOT, etc. at the beginning of each bullet point should be put in capital letters?

File: samples/sample_definition_and_criteria.rst [hunk 8]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    file and should be operational on all supported platforms. Do not rely on the
+    ``extra_args`` or ``extra_configs`` options in the YAML file to build the
+    sample.
+  * The tests should verify the default configuration of the sample and any
+    optional features documented in the sample's README file. Sample should
+    never be used to test functionality of the subsystem or module the sample
+    belongs to.
+  * Any documented configuration options related to the sample and its
+    operation can be verified using ``extra_args`` or
+    ``extra_configs`` options in the YAML file. The `prj.conf` file should have the
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:73]: sphinx-lint will catch this! yay

```suggestion
    ``extra_configs`` options in the YAML file. The :file:`prj.conf` file should have the
```

File: samples/sample_definition_and_criteria.rst [hunk 9]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    base configuration options.
+  * Sample output can be validated leveraging the ``harness_config`` regex option,
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:75]:
```suggestion
  * Sample output can be validated by leveraging the ``harness_config`` regex option,
```

File: samples/sample_definition_and_criteria.rst [hunk 10]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    wherever applicable.
+  * Use ``platform_allow`` to limit test to the platforms the sample was
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:77]:
```suggestion
  * Use ``platform_allow`` to limit test to the platforms the sample was actually
```

File: samples/sample_definition_and_criteria.rst [hunk 11]
```diff
@@ -39,36 +47,42 @@ Sample Criteria
...
+    verified on. Those platforms should be listed in the sample's README file.
   * Make use of ``integration_platforms`` to limit the scope when there are timing or
-    resource constraints.
+    resource constraints. Ideally, one platform should be enough to verify the
+    sample when changes are submitted to the Zephyr tree via a pull-request.
   * Make the sample as generic as possible. Avoid making a sample platform specific unless it is
     for particular hardware.
 
-4. A sample should provide sufficient coverage of a subsystem, feature, or module.
+3. A sample should provide sufficient coverage of a subsystem, feature, or module.
```
Comment (kartben) [inline samples/sample_definition_and_criteria.rst:85]: FWIW I have mixed feelings about this section as it contains a lot of redundant information (if we people don't get it by now that samples must not be unit tests!) and the wording might be confusing: we ask for "sufficient coverage of a subsytesm" when earlier we advise to not use samples to test subsystems.
No immediate action needed though as it is OK enough as-is and can be improved in later PRs.

**Review by kartben — CHANGES_REQUESTED** [review]
thanks for this!

File: samples/sample_definition_and_criteria.rst (hunk 11 above)
Comment (nashif) [reply samples/sample_definition_and_criteria.rst:85]: coverage as in 'feature coverage', try to demonstrate as much as possible of a subsystem to be usable and submit thin content leaving much behind.

Comment (kartben) [reply samples/sample_definition_and_criteria.rst:85]: yup, I know that's what this is *supposed* to mean, hence why I said I think it's *confusing* as other people not aware of the "gating" we're trying to establish might interpret "coverage" in a different way. Like I said, will likley tackle this in a follow-up PR but I guess where my head is at is rewording/summarizing this along the lines of "A sample should demonstrate an actual use case" or something in that vein.

**Review by kartben — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**